src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,217 @@ public static CFProxySettings GetSystemProxySettings ()
 			dict.DangerousRelease ();
 			return new CFProxySettings (dict);
 		}
-		
-#if notyet
-		// FIXME: These require bindings for CFRunLoopSource and CFStreamClientContext
-		
-		public delegate void CFProxyAutoConfigurationResultCallback (IntPtr client, NSArray proxyList, NSError error);
-		
+
+		delegate void CFProxyAutoConfigurationResultCallbackInternal (IntPtr client, IntPtr proxyList, IntPtr error);
+		public delegate void CFProxyAutoConfigurationResultCallback (NSObject client, CFProxy[] proxyList, NSError error);
+		// helper delegate to reuse code
+		delegate IntPtr CreatePACCFRunLoopSource (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context);
+
+		static CFProxy[] ParseProxies (IntPtr proxyList)
+		{
+			CFProxy[] proxies = null;
+			if (proxyList != IntPtr.Zero) {
+				using (var array = new CFArray (proxyList, false)) {
+					proxies = new CFProxy [array.Count];
+					for (int i = 0; i < proxies.Length; i++) {
+						var dict = new NSDictionary (array.GetValue (i));
+						proxies[i] = new CFProxy (dict);
+					}
+				}
+			}
+			return proxies;
+		}
+
+		static async Task<(CFProxy[] proxies, NSError error)> ExecutePacCFRunLoopSourceAsync (CreatePACCFRunLoopSource factory, CancellationToken cancellationToken)
+		{
+			CFProxy[] proxies = null;
+			NSError outError = null;
+			await Task.Run (() => {
```

Comment:
Can you provide more context on why we need to have async calls for this specific case?

Reply:
This implementation is similar to the sync once except that the execution of the **CFRunLoop** run method is executed in  **different thread**. You can think of the CFRunLoop  run method as a while loop that processes messages that are pumped to it via CFRunLoopSource objects. The *run* method will have a while loop that will process the events and will block until stopped.

Moving the run method to a **diff thread** ensures that we do not block the main thread (UI thread most probably) until the PAC script file is executed.

We are providing 3 APIs for this.

1. Sync one that blocks the thread until the PAC is parsed.
2. Sync one that returns the CFRunLoopSource for the user to add to a run loop to me processed.
3. Async one that does all the work for the user so that he does not have to look into the details of getting the correct CFRunLoop and call the run method.

This method is number 3. We ensure that we do not block the current thread and return the result via a task that the user can await.

Reply:
So the pattern is: send off new thread to do something, return a `Task` object that will contain the newly spawned thread's results when it returns?

But if the current thread is awaiting the return from the `Task`, isn't it blocked? Or is there an event that notifies the main thread that the task is complete so that it can do something with the result?

Reply:
So in this case. We crate a new thread that is squealed in the ThreadPool, we set the Task to ConfigureAwait (false) stating that we do not need the sync context for the calling thread and return it.

Because we are not requiring the sync context, the calling thread is not blocked, which is very important in threads that are dealing with events such as the UI thread, and whenever the thread completes, the code of the user will resume executing as if everything was sequential. 

For the point of view of our users, the should write the code as if it was sequential when in reality we are not blocking the UI thread with the CFRunLoop but a diff one.